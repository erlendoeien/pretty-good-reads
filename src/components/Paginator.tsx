import { useReactiveVar } from '@apollo/client';
import { makeStyles } from '@material-ui/core';
import Pagination from '@material-ui/lab/Pagination/Pagination';
// eslint-disable-next-line
import React from 'react';
import { offsetVar, limitVar } from '../cache';

const useStyles = makeStyles((theme) => ({
    root: {
        '& > *': {
            marginTop: theme.spacing(2),
            justifyContent: 'center',
            display: 'flex'
        }
    }
}));

/**
 * Paginator bar, almost independent by reactive variables limit and offset
 */
export default function Paginator({ totalCount }: { totalCount: number }) {
    const offset = useReactiveVar(offsetVar);
    const limit = useReactiveVar(limitVar);
    const classes = useStyles();

    /*
     * Updates the reactive offset to the new value
     */
    const handlePageChange = (_: React.ChangeEvent<unknown>, newPage: number) => {
        offsetVar((newPage - 1) * limit);
    };

    // No books to paginate
    if (totalCount === 0) return null;

    return (
        <Pagination
            className={classes.root}
            page={offset / limit + 1}
            count={Math.ceil(totalCount / limit)}
            variant="outlined"
            shape="rounded"
            onChange={handlePageChange}
        />
    );
}
