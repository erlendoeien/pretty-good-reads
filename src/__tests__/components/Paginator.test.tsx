import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Pagination from '@material-ui/lab/Pagination/Pagination';
import Adapter from 'enzyme-adapter-react-16';
import Paginator from 'src/components/Paginator';

Enzyme.configure({ adapter: new Adapter() });

//snapshot test
describe('testing header', () => {
    it('renders correctly', () => {
        const paginator = shallow(<Paginator totalCount={30} />);
        expect(paginator).toMatchSnapshot();
    });
});

//mock test
const fnChange = jest.fn();
//replacing onChange function with mock function.
describe('change event', () => {
    it('Page change should update offset value', () => {
        const pagination = shallow(<Pagination onChange={fnChange} />);
        pagination.simulate('change');
        expect(fnChange).toHaveBeenCalled();
    });
});
