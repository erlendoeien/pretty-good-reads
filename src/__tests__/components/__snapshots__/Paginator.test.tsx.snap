// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing header renders correctly 1`] = `
<WithStyles(ForwardRef(Pagination))
  className="makeStyles-root-1"
  count={3}
  onChange={[Function]}
  page={1}
  shape="rounded"
  variant="outlined"
/>
`;
